heft-plugins/heft-rspack-plugin/src/RspackPlugin.ts

```diff
@@ -0,0 +1,503 @@
+      throw new Error(
+        `The ${JSON.stringify(
+          SERVE_PARAMETER_LONG_NAME
+        )} parameter is only available when running in watch mode.` +
+          ` Try replacing "${taskSession.parsedCommandLine?.unaliasedCommandName}" with` +
+          ` "${taskSession.parsedCommandLine?.unaliasedCommandName}-watch" in your Heft command line.`
+      );
+    }
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      await this._runRspackAsync(taskSession, heftConfiguration, options);
+    });
+
+    taskSession.hooks.runIncremental.tapPromise(
+      PLUGIN_NAME,
+      async (runOptions: IHeftTaskRunIncrementalHookOptions) => {
+        await this._runRspackWatchAsync(taskSession, heftConfiguration, options, runOptions.requestRun);
+      }
+    );
+  }
+
+  private async _getRspackConfigurationAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IRspackPluginOptions,
+    requestRun?: () => void
+  ): Promise<IRspackConfiguration | undefined> {
+    if (this._rspackConfiguration === false) {
+      const rspackConfiguration: IRspackConfiguration | undefined = await tryLoadRspackConfigurationAsync(
+        {
+          taskSession,
+          heftConfiguration,
+          hooks: this.accessor.hooks,
+          serveMode: this._isServeMode,
+          loadRspackAsyncFn: this._loadRspackAsync.bind(this, taskSession, heftConfiguration)
+        },
+        options
+      );
+
+      this._rspackConfiguration = rspackConfiguration;
+    }
+
+    return this._rspackConfiguration;
+  }
+
+  private async _loadRspackAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration
+  ): Promise<RspackCoreImport> {
+    if (!this._rspack) {
+      try {
+        const rspackPackagePath: string = await heftConfiguration.rigPackageResolver.resolvePackageAsync(
+          RSPACK_PACKAGE_NAME,
+          taskSession.logger.terminal
+        );
+        this._rspack = await import(rspackPackagePath);
+      } catch (e) {
+        // Fallback to bundled version if not found in rig.
+        this._rspack = await import(RSPACK_PACKAGE_NAME);
+        taskSession.logger.terminal.writeDebugLine(`Using Rspack from built-in "${RSPACK_PACKAGE_NAME}"`);
+      }
+    }
+    return this._rspack!;
+  }
+
+  private async _getRspackCompilerAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    rspackConfiguration: IRspackConfiguration
+  ): Promise<TRspack.Compiler | TRspack.MultiCompiler> {
+    if (!this._rspackCompiler) {
+      const rspack: RspackCoreImport = await this._loadRspackAsync(taskSession, heftConfiguration);
+      taskSession.logger.terminal.writeLine(`Using Rspack version ${rspack.version}`);
+      this._rspackCompiler = Array.isArray(rspackConfiguration)
+        ? rspack.default(rspackConfiguration) /* (rspack.Compilation[]) => MultiCompiler */
```

Comment:
I know it is the case in webpack, but does rspack also type this as an overload or is it a function with a union? If the latter you can probably just invoke it directly without the silly ternary

Reply:
looks like it

```ts
declare function createMultiCompiler(options: MultiRspackOptions): MultiCompiler;
declare function createCompiler(userOptions: RspackOptions): Compiler;
declare function rspack(options: MultiRspackOptions): MultiCompiler;
declare function rspack(options: RspackOptions): Compiler;
declare function rspack(options: MultiRspackOptions | RspackOptions): MultiCompiler | Compiler;
declare function rspack(options: MultiRspackOptions, callback?: Callback<Error, MultiStats>): null | MultiCompiler;
declare function rspack(options: RspackOptions, callback?: Callback<Error, Stats>): null | Compiler;
declare function rspack(options: MultiRspackOptions | RspackOptions, callback?: Callback<Error, MultiStats | Stats>): null | MultiCompiler | Compiler;
```